loopy/symbolic.py

```diff
@@ -1587,6 +1587,75 @@ def constraint_to_cond_expr(cns):
 # }}}
 
 
+# {{{ isl_set_from_expr
+
+class AffineConditionToISLSet(IdentityMapper):
+
+    def __init__(self, space):
+        self.space = space
+        super().__init__()
+
+    def map_comparison(self, expr):
+        if expr.operator == "!=":
+            return self.rec(p.LogicalNot(p.Comparison(expr.left, "==", expr.right)))
+
+        left_aff = guarded_aff_from_expr(self.space, expr.left)
+        right_aff = guarded_aff_from_expr(self.space, expr.right)
+
+        if expr.operator == "==":
+            cnst = isl.Constraint.equality_from_aff(left_aff-right_aff)
+        elif expr.operator == ">=":
+            cnst = isl.Constraint.inequality_from_aff(left_aff-right_aff)
+        elif expr.operator == ">":
+            cnst = isl.Constraint.inequality_from_aff(left_aff-right_aff-1)
+        elif expr.operator == "<=":
+            cnst = isl.Constraint.inequality_from_aff(right_aff-left_aff)
+        elif expr.operator == "<":
+            cnst = isl.Constraint.inequality_from_aff(right_aff-left_aff-1)
+        else:
+            assert False
+
+        return isl.BasicSet.universe(self.space).add_constraint(cnst)
+
+    def _map_logical_reduce(self, expr, f):
+        """
+        :arg f: Reduction callable.
+        """
+        sets = [self.rec(child) for child in expr.children]
+
+        if not all(isinstance(set_, (isl.BasicSet, isl.Set)) for set_ in sets):
+            raise LoopyError(f"'{expr}' doesn't have all its children as"
+                    " conditions")
+
+        return reduce(f, sets)
+
+    def map_logical_or(self, expr):
+        import operator
+        return self._map_logical_reduce(expr, operator.or_)
+
+    def map_logical_and(self, expr):
+        import operator
+        return self._map_logical_reduce(expr, operator.and_)
+
+    def map_logical_not(self, expr):
+        set_ = self.rec(expr.child)
+        if not isinstance(set_, (isl.BasicSet, isl.Set)):
+            raise LoopyError(f"'{expr.child}' is not a condition")
+
+        return set_.complement()
+
+
+def isl_set_from_expr(space, expr):
+    mapper = AffineConditionToISLSet(space)
+    set_ = mapper(expr)
+    if not isinstance(set_, (isl.BasicSet, isl.Set)):
+        raise LoopyError(f"'{expr}' is not a condition.")
```

Comment:
I'm confused by this. C-style implicit `!= 0` is OK, FWIW. (Btw we should document that.)

Reply:
Yep, sadly the previous implementation was broken. See ed672ff.

Reply:
Now we also test it. See 0baaaf4.